extmod/machine_i2c_target.c

```diff
@@ -0,0 +1,356 @@
+static void mp_machine_i2c_target_deinit(machine_i2c_target_obj_t *self);
+
+static const mp_irq_methods_t machine_i2c_target_irq_methods;
+
+#define I2C_TARGET_NUM_IRQ (4)
+
+// Needed to retain a root pointer to the memory object.
+MP_REGISTER_ROOT_POINTER(mp_obj_t i2c_target_mem_obj[I2C_TARGET_NUM_IRQ]);
+
+MP_REGISTER_ROOT_POINTER(void *machine_i2c_target_irq_obj[I2C_TARGET_NUM_IRQ]);
+
+void mp_machine_i2c_target_deinit_all(void) {
+    for (size_t i = 0; i < I2C_TARGET_NUM_IRQ; ++i) {
+        MP_STATE_PORT(i2c_target_mem_obj[i]) = MP_OBJ_NULL;
+        MP_STATE_PORT(machine_i2c_target_irq_obj[i]) = NULL;
+    }
+    mp_machine_i2c_target_deinit_all_port();
+}
+
+static bool event(machine_i2c_target_data_t *data, unsigned int trigger, size_t arg0, size_t arg1) {
+    unsigned int id = 0; // TODO
+    machine_i2c_target_irq_obj_t *irq = MP_STATE_PORT(machine_i2c_target_irq_obj[id]);
+    if (irq != NULL && (trigger & irq->trigger)) {
+        irq->flags = trigger & irq->trigger;
+        mp_machine_i2c_target_event_callback(irq);
+        return true; // irq handled
+    }
+    return false; // irq not handled
+}
+
+static void machine_i2c_target_data_init(machine_i2c_target_data_t *data, mp_obj_t mem_obj, mp_int_t mem_addrsize) {
+    data->state = STATE_IDLE;
+    data->mem_addr = 0;
+    data->mem_len = 0;
+    data->mem_buf = NULL;
+    if (mem_obj != mp_const_none) {
+        mp_buffer_info_t bufinfo;
+        mp_get_buffer_raise(mem_obj, &bufinfo, MP_BUFFER_RW);
+        if (mem_addrsize < 8 || mem_addrsize > 32 || mem_addrsize % 8 != 0) {
+            mp_raise_ValueError(MP_ERROR_TEXT("mem_addrsize must be 8, 16, 24 or 32"));
+        }
+        data->mem_len = bufinfo.len;
+        data->mem_buf = bufinfo.buf;
+    }
+}
+
+static void machine_i2c_target_data_reset_helper(machine_i2c_target_data_t *data) {
+    size_t len = 0; // TODO
+    if (data->state == STATE_READING) {
+        event(data, I2C_TARGET_IRQ_READ_END, data->mem_addr, len);
+    } else if (data->state == STATE_WRITING) {
+        event(data, I2C_TARGET_IRQ_WRITE_END, data->mem_addr, len);
+    }
+    data->state = STATE_IDLE;
+}
+
+static void machine_i2c_target_data_addr_match(machine_i2c_target_data_t *data, bool read) {
+    event(data, I2C_TARGET_IRQ_ADDR_MATCH, read, 0);
+    machine_i2c_target_data_reset_helper(data);
+}
+
+static void machine_i2c_target_data_read_request(machine_i2c_target_obj_t *self, machine_i2c_target_data_t *data) {
+    // Let the user handle the read request.
+    bool event_handled = event(data, I2C_TARGET_IRQ_READ_REQ, 0, 0);
+    if (data->mem_buf == NULL) {
+        if (!event_handled) {
+            // No data source, just write out a zero.
+            uint8_t val = 0;
+            mp_machine_i2c_target_write_bytes(self, 1, &val);
+        }
+    } else {
+        // Have a buffer.
+        if (data->state == STATE_MEM_ADDR_SELECTED) {
```

Comment:
The additional calls for ADDR_MATCH events disappear, if the test is:
`        if (data->state == STATE_MEM_ADDR_SELECTED || data->state == STATE_IDLE) {`
This is the smallest change I could find, but may not be the best method to fix the state machine here.
But that seems to be only a problem of the RP2 port, because it does not have a specific ADDR_MATCH event.

Reply:
Just ` if (data->state ==  STATE_IDLE)` is sufficient.